File: src/705. Design HashSet/main.go
package main

type MyHashSet struct {
	set [125001]byte
}

func Constructor() MyHashSet {
	return MyHashSet{}
}

func (this *MyHashSet) Add(key int) {
	byt, bit := this.findPosition(key)
	this.set[byt] |= 1 << bit
}

func (this *MyHashSet) Remove(key int) {
	byt, bit := this.findPosition(key)
	this.set[byt] &= ^(1 << bit)
}

func (this *MyHashSet) Contains(key int) bool {
	byt, bit := this.findPosition(key)
	v := this.set[byt]
	return v&(1<<bit) != 0
}

func (this *MyHashSet) findPosition(key int) (int, int) {
	byt := key / 8
	bit := key % 8

	return byt, bit
}

File: src/705. Design HashSet/main.rs
const DATA_LEN: usize = 125001;

struct MyHashSet {
    data: [usize; DATA_LEN],
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyHashSet {

    fn new() -> Self {
        Self { data: [0; DATA_LEN] } 
    }
    
    fn add(&mut self, key: i32) {
        let (byte, bit) = self.find_position(key);
        self.data[byte] |= (1 << bit);
         
    }
    
    fn remove(&mut self, key: i32) {
        let (byte, bit) = self.find_position(key);
        self.data[byte] &= !(1 << bit);
    }
    
    fn contains(&self, key: i32) -> bool {
        let (byte, bit) = self.find_position(key);

        (self.data[byte] & (1 << bit)) != 0
        
    }

    fn find_position(&self, key: i32) -> (usize, usize) {
        ((key / 8) as usize, (key % 8) as usize)
    }
}

/**
 * Your MyHashSet object will be instantiated and called as such:
 * let obj = MyHashSet::new();
 * obj.add(key);
 * obj.remove(key);
 * let ret_3: bool = obj.contains(key);
 */

